src/target/riscv/riscv-013.c

```diff
@@ -3219,30 +3268,137 @@ static int read_memory(struct target *target, target_addr_t address,
 	if (count == 0)
 		return ERROR_OK;
 
-	RISCV013_INFO(info);
-	if (has_sufficient_progbuf(target, 3) && !riscv_prefer_sba)
-		return read_memory_progbuf(target, address, size, count, buffer,
-			increment);
-
-	if ((get_field(info->sbcs, DM_SBCS_SBACCESS8) && size == 1) ||
-			(get_field(info->sbcs, DM_SBCS_SBACCESS16) && size == 2) ||
-			(get_field(info->sbcs, DM_SBCS_SBACCESS32) && size == 4) ||
-			(get_field(info->sbcs, DM_SBCS_SBACCESS64) && size == 8) ||
-			(get_field(info->sbcs, DM_SBCS_SBACCESS128) && size == 16)) {
-		if (get_field(info->sbcs, DM_SBCS_SBVERSION) == 0)
-			return read_memory_bus_v0(target, address, size, count, buffer,
-				increment);
-		else if (get_field(info->sbcs, DM_SBCS_SBVERSION) == 1)
-			return read_memory_bus_v1(target, address, size, count, buffer,
-				increment);
+	if (size != 1 && size != 2 && size != 4 && size != 8 && size != 16) {
+		LOG_ERROR("BUG: Unsupported size for memory read: %d", size);
+		return ERROR_FAIL;
 	}
 
-	if (has_sufficient_progbuf(target, 3))
-		return read_memory_progbuf(target, address, size, count, buffer,
-			increment);
+	unsigned i;
+	int ret = ERROR_FAIL;
+	bool warn = true;
+	RISCV_INFO(r);
+	RISCV013_INFO(info);
+
+	char *progbuf_result = "disabled";
+	char *sysbus_result = "disabled";
+	char *abstract_result = "disabled";
+
+	for (i = 0; i < RISCV_NUM_MEM_ACCESS_METHODS; i++) {
+		int method = r->mem_access_methods[i];
+
+		if (method == RISCV_MEM_ACCESS_PROGBUF) {
+			if (!has_sufficient_progbuf(target, 3)) {
+				LOG_DEBUG("Skipping mem read via progbuf - insufficient progbuf size.");
+				progbuf_result = "skipped";
+				continue;
+			}
+			if (target->state != TARGET_HALTED) {
+				LOG_DEBUG("Skipping mem read via progbuf - target not halted.");
+				progbuf_result = "skipped";
+				continue;
+			}
+			if (riscv_xlen(target) < size * 8) {
+				LOG_DEBUG("Skipping mem read via progbuf - XLEN (%d) is too short for %d-bit memory read.",
+						riscv_xlen(target), size * 8);
+				progbuf_result = "skipped";
+				continue;
+			}
+			if (size > 8) {
+				LOG_DEBUG("Skipping mem read via progbuf - unsupported size.");
+				progbuf_result = "skipped";
+				continue;
+			}
+			if ((sizeof(address) * 8 > riscv_xlen(target)) && (address >> riscv_xlen(target))) {
+				LOG_DEBUG("Skipping mem write via progbuf - progbuf only supports %u-bit address.",
+						riscv_xlen(target));
+				progbuf_result = "skipped";
+				continue;
+			}
```

Comment:
This looks identical to the code in write_memory_progbuf (except I suppose the read/write strings). Can you refactor this?